build_runner/bin/build_runner.dart

```diff
@@ -56,54 +60,89 @@ Future<Null> main(List<String> args) async {
   if (commandName != _generateCommand) {
     logListener = Logger.root.onRecord.listen(stdIOLogListener());
   }
-  var buildScript = await generateBuildScript();
-  var scriptFile = File(scriptLocation)..createSync(recursive: true);
-  scriptFile.writeAsStringSync(buildScript);
-  if (commandName == _generateCommand) {
-    print(p.absolute(scriptLocation));
-    return;
-  }
 
-  var exitPort = ReceivePort();
-  var errorPort = ReceivePort();
-  var messagePort = ReceivePort();
-  var errorListener = errorPort.listen((e) {
-    stderr.writeln('\n\nYou have hit a bug in build_runner');
-    stderr.writeln('Please file an issue with reproduction steps at '
-        'https://github.com/dart-lang/build/issues\n\n');
-    final error = e[0];
-    final trace = e[1] as String;
-    stderr.writeln(error);
-    stderr.writeln(Trace.parse(trace).terse);
-    if (exitCode == 0) exitCode = 1;
-  });
-  try {
-    await Isolate.spawnUri(
-        Uri.file(p.absolute(scriptLocation)), args, messagePort.sendPort,
-        onExit: exitPort.sendPort, onError: errorPort.sendPort);
-  } on IsolateSpawnException catch (e) {
-    print(red.wrap('Failed to launch the build script. '
-        'This is likely due to a misconfigured builder definition. '
-        'See the generated script at $scriptLocation to find errors.'));
-    print(e);
-    messagePort.sendPort.send(ExitCode.config.code);
-    exitPort.sendPort.send(null);
-  }
-  StreamSubscription exitCodeListener;
-  exitCodeListener = messagePort.listen((isolateExitCode) {
-    if (isolateExitCode is! int) {
-      throw StateError(
-          'Bad response from isolate, expected an exit code but got '
-          '$isolateExitCode');
+  var shouldRerun = true;
+  while (shouldRerun) {
```

Comment:
what do you think about factoring this out into a method that returns the exit code from the isolate - then the loop can be `while (await generateAndRun() == ExitCode.tempFail.code);`